Issue/3192 pydantic anyurl doc rendering

edvgui opened the pull request:
# Description

Fixing formating of AnyUrl types in the documentation (openapi and swagger).

closes #3192 and inmanta/inmanta-lsm#451

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

wouterdb reviewed (commented):
What strikes me as fundamentally odd it that we serialize AnyUrl in two different ways:
1. if it is inside a pydantic object, we use the pydantic model_schema method
2. if it top level, we use our own logic

This is a pre-existing condition, but I wonder is we could not simple get rid of 99% of the logic here by simple making a method into a pydantic object, get the schema out,...

Give then duplication, it is important that we stay aligned. I added the following to the test  and it confirms that it is serialized consistently. 

```python
def test_openapi_types_anyurl():
    type_converter = OpenApiTypeConverter()

    def conformance_test(openapi_type, the_type):
        class Sub(BaseModel):
            the_field: the_type

        pydantic_result = type_converter._handle_pydantic_model(Sub).properties["the_field"]
        pydantic_result.title = None
        assert openapi_type == pydantic_result

    openapi_type = type_converter.get_openapi_type(AnyUrl)
    assert openapi_type == Schema(type="string", format="uri")
    conformance_test(openapi_type, AnyUrl)

    openapi_type = type_converter.get_openapi_type(AnyHttpUrl)
    assert openapi_type == Schema(type="string", format="uri")
    conformance_test(openapi_type, AnyHttpUrl)

    openapi_type = type_converter.get_openapi_type(PostgresDsn)
    assert openapi_type == Schema(type="string", format="uri")
    conformance_test(openapi_type, PostgresDsn)
```

I can push this change if you want to.

bartv commented:
@edvgui I removed myself as reviewer because I really do not know the code. Can you explain why you changed the `elif` to `if`? This is equivalent but it seems unnecessary to make this change. Unless I am missing something.

edvgui commented:
> @edvgui I removed myself as reviewer because I really do not know the code. Can you explain why you changed the `elif` to `if`? This is equivalent but it seems unnecessary to make this change. Unless I am missing something.

I think it looks better/cleaner/lighter, but this is definitely not worth fighting over.  I changed it back.

edvgui commented on tests/test_openapi.py:
```diff
@@ -343,6 +345,30 @@ def test_openapi_types_uuid():
+def test_openapi_types_anyurl():
+    type_converter = OpenApiTypeConverter()
+
+    def conformance_test(openapi_type: Schema, the_type: Type):
```
@wouterdb I added your suggested changes, but I have to say that I don't understand what it does.  Could you explain a bit further?

I don't understand either where is the duplicated logic

edvgui replied:
Also, the test now fails with this error:
```
>       assert openapi_type == pydantic_result
E       AssertionError: assert Schema(ref=None, title=None, required=None, type='string', items=None, properties=None, description=None, format='uri', default=None, nullable=None, readOnly=None, example=None, deprecated=None, anyOf=None, enum=None, additionalProperties=None) == SchemaBase(ref=None, title=None, required=None, type='string', items=None, properties=None, description=None, format='uri', default=None, nullable=None, readOnly=None, example=None, deprecated=None, anyOf=None, enum=None)
E         +Schema(ref=None, title=None, required=None, type='string', items=None, properties=None, description=None, format='uri', default=None, nullable=None, readOnly=None, example=None, deprecated=None, anyOf=None, enum=None, additionalProperties=None)
E         -SchemaBase(ref=None, title=None, required=None, type='string', items=None, properties=None, description=None, format='uri', default=None, nullable=None, readOnly=None, example=None, deprecated=None, anyOf=None, enum=None)
E         Full diff:
E         - SchemaBase(ref=None, title=None, required=None, type='string', items=None, properties=None, description=None, format='uri', default=None, nullable=None, readOnly=None, example=None, deprecated=None, anyOf=None, enum=None)
E         ?       ----
E         + Schema(ref=None, title=None, required=None, type='string', items=None, properties=None, description=None, format='uri', default=None, nullable=None, readOnly=None, example=None, deprecated=None, anyOf=None, enum=None, additionalProperties=None)
E         ?
```
But I don't know if it means my change created consistency issues or if the test is wrong?

wouterdb reviewed (commented):
As discussed, you can re-request review when the next iteration is done

inmantaci commented:
Processing this pull request

inmantaci commented:
Failed to merge changes into iso4 due to merge conflict.Please open a pull request for these branches separately.

inmantaci commented:
Merged into branches master in 3bd93bc25fdd017b6e63da43534a244d4b14babc

inmantaci commented:
Not closing this pull request due to previously commented issues for some of the destination branches. Please open a separate pull request for those branches. You can safely close this pull request and delete the source branch.
